fixed bug in DataFrame.diff - issue #10907

terrytangyuan opened the pull request:
Fixed issue #10907

terrytangyuan committed:
BUG: Fixed bug in DataFrame.diff for issue 10907 when DataFrame is not consolidated (+1 squashed commit)
Squashed commits:
[6fe71d3] moved changes to correct place and fixed test_diff (+1 squashed commit)
Squashed commits:
[2bf3c2b] moved change to where diff is defined and updated test (+1 squashed commit)
Squashed commits:
[6715d7f] added unit test to test this fix (+1 squashed commit)
Squashed commits:
[f06fa5e] fixed bug in DataFrame.diff

terrytangyuan commented:
This is my first PR here. Could someone help me pass this Travis test? After I made the changes in the code, what should I do?

jreback commented:
try running the tests locally first 

eg

nosetests pandas/test_frame.py

jreback commented:
the consolidate has to go in core/internals.py (where diff is defined)

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,11 @@ def test_diff(self):
 
+        df = pd.DataFrame({'y': pd.Series([2]), 'z': pd.Series([3])})
+        df.insert(0, 'x', 1)
+        self.assertEqual(df.diff(axis=1)['x']==np.nan, False)
```
construct an expected frame the use
assert_frame_equal

terrytangyuan commented:
When I run tests by calling `nosetests pandas/tests/test_frame.py`, the following error occurred: `ImportError: No module named test_frame`. 

Is there anything else I should do before test it?

jreback commented:
see the docs here: http://pandas.pydata.org/pandas-docs/stable/contributing.html

jreback commented on pandas/core/internals.py:
```diff
@@ -2513,7 +2513,7 @@ def putmask(self, **kwargs):
 
     def diff(self, **kwargs):
-        return self.apply('diff', **kwargs)
+        return self.consolidate().apply('diff', **kwargs)
```
instead of this

add to ``.apply`` an argument ``consolidate=True``, and:
```
if consolidate:
    self._consolidate_inplace()
```
This should fix this issue and not break anything else.

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,13 @@ def test_diff(self):
 
+        df = pd.DataFrame({'y': pd.Series([2]), 'z': pd.Series([3])})
+        df.insert(0, 'x', 1)
+        the_diff = df.consolidate().diff(axis=1)
```
just do

```
result = df.diff(axis=1)
expected = DataFrame(....)
assert_frame_equal(result, expected)
```

terrytangyuan commented:
Thanks so much for your help! I have made the changes. Now I understand the code structure better. Hopefully I'll be able to contribute more in the future.

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,12 @@ def test_diff(self):
                             tf['A'] - tf['A'].shift(1))
 
+        df = pd.DataFrame({'x': pd.Series([1]),'y': pd.Series([2]), 'z': pd.Series([3])})
+        result = df.diff(axis=1).astype(float)
```
don't modify result 

the astype might be hiding something

jreback commented on pandas/core/internals.py:
```diff
@@ -2414,7 +2414,7 @@ def _verify_integrity(self):
 
-    def apply(self, f, axes=None, filter=None, do_integrity_check=False, **kwargs):
+    def apply(self, f, axes=None, filter=None, do_integrity_check=False, consolidate=True, **kwargs):
         """
```
can u update the doc-string (u may have to add the other parms as well)

terrytangyuan committed:
DOC: Added a note in whatsnew and doc-string for fixing issue 10907  (+1 squashed commit)
Squashed commits:
[810cbda] DOC: Added a note in whatsnew and doc-string for fixing issue 10907 (+1 squashed commit)
Squashed commits:
[f9220a2] DOC: Added a note in whatsnew and doc-string for fixing issue 10907 (+1 squashed commit)
Squashed commits:
[0f1836f] added consolidate as param in doc-string

terrytangyuan commented:
Done. Should be able to merge now since it passes tests locally. There are too many Travis build tasks pending now.

jreback commented:
needs a whatsnew note
pls squash

terrytangyuan commented:
This should do it now, right?

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,12 @@ def test_diff(self):
                             tf['A'] - tf['A'].shift(1))
 
+        df = pd.DataFrame({'x': pd.Series([1]),'y': pd.Series([2]), 'z': pd.Series([3])})
+        result = df.diff(axis=1)
```
add the issue number as a comment here.

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,12 @@ def test_diff(self):
+        df = pd.DataFrame({'x': pd.Series([1]),'y': pd.Series([2]), 'z': pd.Series([3])})
+        result = df.diff(axis=1)
+        expected = pd.DataFrame({'x':np.nan, 'y':pd.Series(1), 'z':pd.Series(1)}).astype(float)
+        self.assert_frame_equal(result, expected)
```
use ``astype('float64')``

jreback commented:
pls squash as well.

terrytangyuan commented:
Thanks. Done.

jreback commented:
ok, pls squash to a single commit. ping when green.

terrytangyuan commented:
It has conflicts that I don't know how to solve if I squash...could we leave it as it is this time? Both commits are descriptive to me.

jreback commented:
ping me when green. typically we use just a single commit on things like this. this is just the way pandas has always worked.

terrytangyuan commented:
@jreback could you merge?

jreback commented:
will get to soon

terrytangyuan commented:
@jreback in case you forgot.

jreback commented on pandas/tests/test_frame.py:
```diff
@@ -10771,6 +10771,13 @@ def test_diff(self):
                             tf['A'] - tf['A'].shift(1))
 
+        # issue 10907
+        df = pd.DataFrame({'x': pd.Series([1]),'y': pd.Series([2]), 'z': pd.Series([3])})
```
this is doesn't actually test anything (and it works), but it doesn't replicate the original issue problem. IIRC you had that their, but it has changed.

terrytangyuan commented:
This is the same as the original. Now in the testing code, I've created the same DataFrame that's originally created via DataFrame.insert. The core problem rises from DataFrame.diff instead of DataFrame.insert. Please take a quick comparison between the outdated diff in this PR and the new push in this PR. Thanks.

terrytangyuan commented:
If you do a assertDataFrameEqual to the original test code and the current test code. You'll see those DataFrames are equivalent.

jreback commented:
your test passes on 0.16.2
so you have not replicated it exactly

you haven't proven that you fixed the bug (you did) but your test doesn't work

terrytangyuan commented:
Ok didn't know that. I'll revise the test soon. Thanks for double checking.
On Aug 31, 2015 3:44 PM, "Jeff Reback" <notifications@github.com> wrote:

> your test passes on 0.16.2
> so you have not replicated it exactly
>
> you haven't proven that you fixed the bug (you did) but your test doesn't
> work
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/10930#issuecomment-136479652>.
>

terrytangyuan commented:
@jreback Could you double check new push?

jreback commented:
merged via https://github.com/pydata/pandas/commit/219ad757a36f96110810282257e3cf2633f1f5bb

thanks!

generally you want to create a branch each time you make a change (and not do it on master)

see [here](http://pandas.pydata.org/pandas-docs/stable/contributing.html)

hhuuggoo commented:
@jreback I don't think this fix works if you have mixed types, since consolidate can't consolidate those blocks.  This is much less common, but I could see someone with mixed ints and floats that wants to shift

jreback commented:
@hhuuggoo yes, this is prob just wrong. it needs to create the indexer then do a take.

Can you create another issue with an example?

thxs

hhuuggoo commented:
yup!
